stake-pool/single-pool/src/processor.rs

```diff
@@ -1029,3 +1029,323 @@ impl Processor {
+
+            // generate between 1 and 100 users and have ~half of them deposit
+            // note for most of these tests we adhere to the minimum delegation
+            // one of the thing we want to see is deposit size being many ooms larger than reward size
+            let mut users = vec![];
+            let user_count: usize = prng.gen_range(1..=100);
+            for _ in 0..user_count {
+                let user = Keypair::new().pubkey();
+
+                if prng.gen_bool(0.5) {
+                    pool.deposit(&user, deposit_amount(&mut prng)).unwrap();
+                }
+
+                users.push(user);
+            }
+
+            // now we do a set of arbitrary operations and confirm invariants hold
+            // we underweight withdraw a little bit to lessen the chances we random walk to an empty pool
+            for _ in 0..1000 {
+                match op_range.sample(&mut prng) {
+                    // deposit a random amount of stake for tokens with a random user
+                    // check their stake, tokens, and share increase by the expected amount
+                    n if n <= 0.35 => {
+                        let user = users.choose(&mut prng).unwrap();
+                        let prev_share = pool.share(user);
+                        let prev_stake = pool.stake(user);
+                        let prev_token_supply = pool.token_supply;
+                        let prev_total_stake = pool.total_stake;
+
+                        let stake_deposited = deposit_amount(&mut prng);
+                        let tokens_minted = pool.deposit(user, stake_deposited).unwrap();
+
+                        // stake increased by exactly the deposit amount
+                        assert_eq!(pool.total_stake - prev_total_stake, stake_deposited);
+
+                        // calculated stake fraction is within 2 lamps of deposit amount
+                        // XXX note to jon: off-by-ones to determine are benign
+                        assert!(
+                            (pool.stake(user) as i64 - prev_stake as i64 - stake_deposited as i64)
+                                .abs()
+                                <= 2
+                        );
+
+                        // tokens increased by exactly the mint amount
+                        assert_eq!(pool.token_supply - prev_token_supply, tokens_minted);
+
+                        // tokens per supply increased with stake per total
+                        if prev_total_stake > 0 {
+                            assert_eq!(
+                                format!("{:.6}", pool.share(user) - prev_share),
+                                format!(
+                                    "{:.6}",
+                                    pool.stake(user) as f64 / pool.total_stake as f64
+                                        - prev_stake as f64 / prev_total_stake as f64
+                                )
+                            );
```

Comment:
because to my great shame i got an error like this
```
error[E0283]: type annotations needed for `Uniform<{float}>`
    --> stake-pool/single-pool/src/processor.rs:1193:24
     |
1193 |         let op_range = Uniform::from(if with_rewards { 0.0..1.0 } else { 0.0..0.65 });
     |             --------   ^^^^^^^ cannot infer type for type `{float}`
     |             |
     |             consider giving `op_range` the explicit type `Uniform<{float}>`, where the type parameter `{float}` is specified
     |
     = note: multiple `impl`s satisfying `{float}: SampleUniform` found in the `rand` crate:
             - impl SampleUniform for f32;
             - impl SampleUniform for f64;
note: required by a bound in `Uniform`
    --> /home/hana/.cargo/registry/src/github.com-1ecc6299db9ec823/rand-0.8.5/src/distributions/uniform.rs:179:23
     |
179  | pub struct Uniform<X: SampleUniform>(X::Sampler);
     |                       ^^^^^^^^^^^^^ required by this bound in `Uniform`
help: consider specifying the type argument in the function call
     |
1193 |         let op_range = Uniform::<X>::from(if with_rewards { 0.0..1.0 } else { 0.0..0.65 });
     |                               +++++
```

and figured i was doing something wrong with the macro, without realizing it was just that i was passing an int for a float...